File: backends/Contracts/GeoRefDataService.Default/Handlers/ReqUpdate.cs
// -----------------------------------------------------------------------
// <copyright file="ReqUpdate.cs" company="DCSV">
// Copyright (c) DCSV. All rights reserved.
// </copyright>
// -----------------------------------------------------------------------

namespace D2.Contracts.GeoRefDataService.Default.Handlers;

using System.Net;
using D2.Contracts.Handler;
using D2.Contracts.Result;
using D2.Services.Protos.Geo.V1;
using Grpc.Core;
using Microsoft.Extensions.Logging;
using H = D2.Contracts.Interfaces.GeoRefDataService.IGeoRefDataService.IReqUpdateHandler;
using I = D2.Contracts.Interfaces.GeoRefDataService.IGeoRefDataService.ReqUpdateInput;
using O = D2.Contracts.Interfaces.GeoRefDataService.IGeoRefDataService.ReqUpdateOutput;

/// <summary>
/// Handler for requesting a reference data update from the Geo service.
/// </summary>
public class ReqUpdate : BaseHandler<ReqUpdate, I, O>, H
{
    private readonly GeoService.GeoServiceClient r_geoClient;

    /// <summary>
    /// Initializes a new instance of the <see cref="ReqUpdate"/> class.
    /// </summary>
    ///
    /// <param name="geoClient">
    /// The Geo service gRPC client.
    /// </param>
    /// <param name="context">
    /// The handler context.
    /// </param>
    public ReqUpdate(
        GeoService.GeoServiceClient geoClient,
        IHandlerContext context)
        : base(context)
    {
        r_geoClient = geoClient;
    }

    /// <summary>
    /// Executes the handler to request a reference data update.
    /// </summary>
    ///
    /// <param name="input">
    /// The input parameters for the handler.
    /// </param>
    /// <param name="ct">
    /// The cancellation token.
    /// </param>
    ///
    /// <returns>
    /// The result of the request operation.
    /// </returns>
    protected override async ValueTask<D2Result<O?>> ExecuteAsync(
        I input,
        CancellationToken ct = default)
    {
        try
        {
            var response = await r_geoClient.RequestReferenceDataUpdateAsync(
                new RequestReferenceDataUpdateRequest(),
                cancellationToken: ct);

            return response.Accepted
                ? D2Result<O?>.Ok(new O(), traceId: TraceId)
                : D2Result<O?>.Fail(
                    ["Update request was not accepted."],
                    HttpStatusCode.ServiceUnavailable,
                    errorCode: ErrorCodes.SERVICE_UNAVAILABLE,
                    traceId: TraceId);
        }
        catch (RpcException ex)
        {
            Context.Logger.LogError(
                ex,
                "RpcException occurred while requesting reference data update. TraceId: {TraceId}",
                TraceId);

            return D2Result<O?>.Fail(
                ["Unable to reach Geo service."],
                HttpStatusCode.ServiceUnavailable,
                errorCode: ErrorCodes.SERVICE_UNAVAILABLE,
                traceId: TraceId);
        }
    }
}

File: backends/node/services/geo/geo-client/src/handlers/c/req-update.ts
import { BaseHandler, type IHandlerContext } from "@d2/handler";
import { D2Result } from "@d2/result";
import { handleGrpcCall } from "@d2/result-extensions";
import type { GeoServiceClient, RequestReferenceDataUpdateResponse } from "@d2/protos";
import type { Commands } from "../../interfaces/index.js";

type Input = Commands.ReqUpdateInput;
type Output = Commands.ReqUpdateOutput;

/**
 * Handler for requesting a reference data update from the Geo service via gRPC.
 * Mirrors D2.Geo.Client.CQRS.Handlers.C.ReqUpdate in .NET.
 */
export class ReqUpdate
  extends BaseHandler<Input, Output>
  implements Commands.IReqUpdateHandler
{
  private readonly geoClient: GeoServiceClient;

  constructor(geoClient: GeoServiceClient, context: IHandlerContext) {
    super(context);
    this.geoClient = geoClient;
  }

  protected async executeAsync(_input: Input): Promise<D2Result<Output | undefined>> {
    const r = await handleGrpcCall(
      () =>
        new Promise<RequestReferenceDataUpdateResponse>((resolve, reject) => {
          this.geoClient.requestReferenceDataUpdate({}, (err, res) => {
            if (err) reject(err);
            else resolve(res);
          });
        }),
      (res) => res.result!,
      (res) => ({ version: res.data?.version }),
    );

    return D2Result.bubble(r, r.data);
  }
}

export type { ReqUpdateInput, ReqUpdateOutput } from "../../interfaces/c/req-update.js";
